### PR TITLE
Update Gogs

### PR DESCRIPTION
**Pull request by Josue-T**
Add : 
- ldap support (for all user + admin)
- multiarch support (build on the script)
- public fonction (creating a new user is blocked per default)

Upgrade script

For ldap support you must reinstall gogs or configure manually ldap in mysql database (the upgrade script don't add ldap support).

### CODE REVIEW

**Comment by stevenroose** [comment]
Nice, this is exactly what I've been trying to add, but I couldnt find the time. Great work!!

**Comment by stevenroose** [comment]
Did you also add all the appropriate ldap fields for automatic Surname, First Name and E-mail data for the users?

The input field allows it (http://gogs.io/docs/features/authentication.html), but I had difficulty to find the accompanying SQL data to put the LDAP fields in...

**Comment by Josue-T** [comment]
For E-mail and username it's ok, but for "Full name" you have to change this parameter in gogs.

**Comment by stevenroose** [comment]
I found the attributes for the Surname and Firstname:
Surname is `sn` and First name is `givenName`.

They can be easily added to the SQL file for the LDAP config.

This is the data in JXplorer:

**Comment by Josue-T** [comment]
I correct this thing.

**Comment by mbugeia** [comment]
Hi, again thanks you very much for this, great work! I did some test and unfortunatly after install I can't login neither with admin account or LDAP account. 

I've look in the database and it seems that the admin user was not created. I don't really know what went wrong, maybe curl, I need to run some other test to find the bug.

I've try both x64 and other architecture install, both installation seems to work fine (but I still can't login). Since you did the work for building gogs from source, what do you think about droping the package.io support? I think it's more convenient to have only one way to install/update/remove gogs and the package doesn't add any value here.

Another thing, with current behavior the script build the development version of Gogs, I think it's a little risky and it should build a specific release.

**Comment by stevenroose** [comment]
Why would package.io support be dropped? The gogs package is actively
maintained by the gogs main dev, right? I think being able to update gogs
with a simple sudo apt-get upgrade is very useful.

I didn't have the chance to test the update yet, but hopefully the source
of the issue can be found and fixed quickly!

**Comment by mbugeia** [comment]
Well I can think of several reason to do so.

A yunohost package is usually consists of install of the app + nginx config + app config. That's mean the package is configurated to work with a specific version of the app. Gogs's packager.io package only take care of the installation. Therefore upgrading Gogs with the package can lead to inconsistency between the installed Gogs's version and nginx and/or yunohost config.
An outdated version that works is better than an up to date one that don't ;)

Moreover the package is only available for x64 architecture so keeping the support mean keeping two way of installing, upgrading, removing the app. So it helps keeping the yunohost package simple and easier to maintain.

Finally the yunohost way of upgrading apps is "yunohost app upgrade appname" and I think it's better to keep it this way. With that, the app is only upgraded if the package have been upgraded and tested with the new version.

tl;dr: consistency and maintainability

**Comment by stevenroose** [comment]
Hmm, those are valid arguments I did not yet consider. 

I must say they made me be in favor of installing from source as well. Isn't it possible to specify a git tag in the install and upgrade scripts from which to install? That way, always the currect version of the app is installed.

One drawback of this method is that every gogs update requires a gogs_ynh update. Perhaps it's possible to somehow allow upgrades within the same `vX.*` branch.

**Comment by Josue-T** [comment]
An other solution is to use only the source for all arch (we don't use the binary for x86-64). And we put the gogs's sources in sources directory (in package) and we build gogs from the sources directory.
With this solution we don't have to support 2 solution (binary or sources).

For the ldap bug, I have installed gogs on a vm and my production and ldap run.
But I have found this commit, https://github.com/gogits/gogs/commit/6969c20afd136ac911c66fd7c7ed88d611ba79fc and this https://github.com/gogits/gogs/commit/cd37fccdfbdf5a1a5b2d85263ffb219068d19205

Today, I have tried an installation and the `allow_auto_register` collum don't exist in mysql database but on last friday it existed.
It's a good example how we have a problem with file how don't came for the package.

**Comment by Josue-T** [comment]
Hello,

I have updated with app. Actually I use only the sources and it build gogs for all arch. The gogs's sources are stored in this package.

**Comment by stevenroose** [comment]
Did you try a fresh install on your machine? Did it work? Can we get this merged? :D

**Comment by Josue-T** [comment]
I have tried on a vm. But on my production it also run. I think it could be merged.

**Comment by stevenroose** [comment]
What is keeping you from merging, @mbugeia ?

(I don't want to push you to just merge without verifying, but perhaps there is something not clear that we can help figure out.)

**Comment by mbugeia** [comment]
Thanks for remind me this. So I tried to install it but it does not work on my yunohost (debian jessie, x64).

It seems that this line https://github.com/Josue-T/gogs_ynh/blob/master/scripts/install#L111 fail silently so everything after that fail too. I can't find any log... MySQL table are not created so the curl command fail.
Additional question, for Jessie it should not rather be 'sudo systemctl start gogs.service' ?

**Comment by Josue-T** [comment]
Can I see the log of your install because I had tried on vm with debian jessie and debian wheezy.
I think it could be a problem with buid but I not sure.

You can try to lunch gogs manually with this command : 

```
export USER=gogs
export HOME=/home/gogs
cd
sudo -u gogs /opt/gogs/gogs web
```

For log on debian jessie you can see log in /var/log/syslog
I think `sudo service gogs start` is same as  `sudo systemctl start gogs.service`

**Comment by mbugeia** [comment]
Hum, sorry the issue was on my side, I don't know exactly what went wrong but after some cleanup and a reboot everything seems to work know :)

I will merge this, thanks you again for your work!
